File: app/k9mail/src/test/java/com/fsck/k9/MockHelper.java
package com.fsck.k9;


import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import static org.mockito.Mockito.RETURNS_DEFAULTS;
import static org.mockito.Mockito.mock;


public class MockHelper {
    public static <T> T mockBuilder(Class<T> classToMock) {
        return mock(classToMock, new Answer() {
            @Override
            public Object answer(InvocationOnMock invocation) throws Throwable {
                Object mock = invocation.getMock();
                if (invocation.getMethod().getReturnType().isInstance(mock)) {
                    return mock;
                } else {
                    return RETURNS_DEFAULTS.answer(invocation);
                }
            }
        });
    }
}

File: legacy/testing/src/main/java/com/fsck/k9/testing/MockHelper.kt
package com.fsck.k9.testing

import org.mockito.Mockito
import org.mockito.Mockito.mock
import org.mockito.kotlin.KStubbing

object MockHelper {
    @JvmStatic
    fun <T> mockBuilder(classToMock: Class<T>): T {
        return mock(classToMock) { invocation ->
            val mock = invocation.mock
            if (invocation.method.returnType.isInstance(mock)) {
                mock
            } else {
                Mockito.RETURNS_DEFAULTS.answer(invocation)
            }
        }
    }

    inline fun <reified T : Any> mockBuilder(stubbing: KStubbing<T>.(T) -> Unit = {}): T {
        return mockBuilder(T::class.java).apply { KStubbing(this).stubbing(this) }
    }
}
